ton/wallet/proof.go

```diff
@@ -114,45 +122,84 @@ func buildMessage(addr *address.Address, proof TonConnectProof) ([]byte, error)
-	key, err := GetPublicKey(ctx, v.client, addr)
+	// First, try to get public key "offline" from stateInit
+	key, err := getPublicKeyFromStateInit(addr, stateInit)
+
+	// Only if it's impossible to parse public key from stateInit, make a request to get it
 	if err != nil {
-		var cErr ton.ContractExecError
-		if !errors.As(err, &cErr) || cErr.Code != ton.ErrCodeContractNotInitialized {
+		key, err = GetPublicKey(ctx, v.client, addr)
+		if err != nil {
 			return nil, fmt.Errorf("failed to get public key: %w", err)
 		}
+	}
 
-		if len(stateInit) == 0 {
-			return nil, errors.New("wallet is not initialized and state init is empty")
-		}
+	return key, nil
+}
 
-		siCell, err := cell.FromBOC(stateInit)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse state init boc: %w", err)
-		}
+func getPublicKeyFromStateInit(addr *address.Address, stateInit []byte) (ed25519.PublicKey, error) {
+	if len(stateInit) == 0 {
+		return nil, errors.New("wallet is not initialized and state init is empty")
+	}
 
-		if !bytes.Equal(siCell.Hash(), addr.Data()) {
-			return nil, errors.New("state init hash does not match address")
-		}
+	siCell, err := cell.FromBOC(stateInit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse state init boc: %w", err)
+	}
 
-		var si tlb.StateInit
-		if err = tlb.LoadFromCell(&si, siCell.BeginParse()); err != nil {
-			return nil, fmt.Errorf("failed to parse state init: %w", err)
-		}
+	if !bytes.Equal(siCell.Hash(), addr.Data()) {
+		return nil, errors.New("state init hash does not match address")
+	}
 
-		if si.Code == nil || si.Data == nil {
-			return nil, errors.New("state init has no code or data")
-		}
+	var si tlb.StateInit
+	if err = tlb.LoadFromCell(&si, siCell.BeginParse()); err != nil {
+		return nil, fmt.Errorf("failed to parse state init: %w", err)
+	}
 
-		ver, ok := walletVersionByCodeHash[string(si.Code.Hash())]
-		if !ok {
-			return nil, errors.New("state init has unknown code")
-		}
+	if si.Code == nil || si.Data == nil {
+		return nil, errors.New("state init has no code or data")
+	}
 
-		key, err = ParsePubKeyFromData(ver, si.Data)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse public key from code: %w", err)
-		}
+	ver, ok := walletVersionByCodeHash[string(si.Code.Hash())]
+	if !ok {
+		return nil, errors.New("state init has unknown code")
 	}
 
+	key, err := ParsePubKeyFromData(ver, si.Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public key from code: %w", err)
+	}
 	return key, nil
 }
+
+func GeneratePayload(secret string, ttl time.Duration) (string, error) {
+	payload := make([]byte, 16, 48)
+	_, err := rand.Read(payload[:8])
+	if err != nil {
+		return "", fmt.Errorf("could not generate nonce")
+	}
+	binary.BigEndian.PutUint64(payload[8:16], uint64(time.Now().Add(ttl).Unix()))
+	h := hmac.New(sha256.New, []byte(secret))
+	h.Write(payload)
+	payload = h.Sum(payload)
+	return hex.EncodeToString(payload[:32]), nil
```

Comment:
Lets increase random size to 24 bytes and keep whole hash, not only 16 bytes, it costs us almost nothing, but looks more secure

Reply:
While hashing only 16 bytes we can get only 2^128 unique hashes (so it is almost same as keeping 16 bytes of hash), but when we increasing payload to 32 bytes we could get up to 2^256 unique hashes, and full hash security is utilized. 

So i think lets increase random size and check full hash